sklearn/datasets/value_dropper.py

```diff
@@ -0,0 +1,401 @@
+                 missing_distribution=None, copy=True, random_state=None):
+        self.missing_values = missing_values
+        self.missing_distribution = missing_distribution
+        self.copy = copy
+        self.random_state = random_state
+
+    def transform(self, X, y=None):
+        """Drop values from ``X`` according to the given distribution.
+
+        Parameters
+        ----------
+
+        X : ndarray like of shape (n_features, n_samples)
+            Data, in which the values must be dropped and set to
+            ``missing_values``.
+
+        y : array-like, shape = (n_samples,), optional for MCAR
+            Target relative to X for classification or regression;
+            When missing_distribution is not a dict (for MCAR missingness),
+            ``y`` need not be passed.
+        """
+        # Validate missing_values and generate missing_mask
+        if ((isinstance(self.missing_values, str) and
+                (self.missing_values.lower() == "nan")) or
+                np.isnan(self.missing_values)):
+            missing_values = np.nan
+            missing_mask = np.isnan(X)
+        else:
+            missing_values = self.missing_values
+            missing_mask = X == self.missing_values
+
+        # Don't allow pre-exising missing values in X, to simplify API
+        X = check_array(X, allow_nan=False,
+                        dtype=('numeric'
+                               if isinstance(self.missing_values,
+                                             (numbers.Integral, np.integer))
+                               else np.float),
+                        copy=self.copy)
+
+        # Validate y, and find type of missingness
+        if isinstance(self.missing_distribution, dict):
+            missing_type = 'nmar'
+            if y is None:
+                raise ValueError("The missing_distribution is a dict "
+                                 "but y is None. If missingness is to be "
+                                 "related to the class labels, target class "
+                                 "labels (y) must be passed.")
+
+            target_type = type_of_target(y)
+            if 'continuous' in target_type or 'multioutput' in target_type:
+                raise ValueError("Value dropping based on the given "
+                                 "distribution can be done only for single "
+                                 "target which is discrete (classification "
+                                 "tasks). The given target (y) is of type %s"
+                                 % target_type)
+            y = check_array(y, ensure_2d=False, dtype='numeric')
+        else:
+            missing_type = 'mcar'
+
+        n_samples, n_features = X.shape
+        n_values = n_samples * n_features
+
+        rng = check_random_state(self.random_state)
+
+        # For NMAR
+        # Convert missing_distribution dict to probability matrix
+        if missing_type == 'nmar':
```

Comment:
Ah sure sorry. I stupidly assumed you asked me to merge this one with the next MCAR / NMAR branching. I'll do this in a minute! Thx